### DISCUSSION TITLE
await Avalonia.Threading.Dispatcher.UIThread

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#14192] [Ideas]
## Context
While i develop internal Visual Studio extension i found this helpfull extension:

```c#

 await Package.JoinableTaskFactory.SwitchToMainThreadAsync(); //  Switch to  Main Thread;

// Execute instruction on Main Thread
// ...

await TaskScheduler.Default; // Switch on  TaskScheduler context

// Execute instruction on TaskScheduler Thread
// ...
``` 

do similar behavior in  `Avalonia 11.0.6`

```c#
await Avalonia.Threading.Dispatcher.UIThread.AwaitWithPriority(Task.Delay(1), Avalonia.Threading.DispatcherPriority.Normal); //  Switch to  Main Thread;

// Execute instruction on UI Thread
// ...

await Task.Factory.StartNew(()=>
{
  // Execute instruction on TaskScheduler Thread context
},
 default,
 TaskCreationOptions.None,
 TaskScheduler.Default)

``` 

## Proposal

Add following API

- public DispatcherPriorityAwaitable AwaitWithPriority(DispatcherPriority priority)
- make Avalonia.Threading.Dispatcher.UIThread awaitable
- make TaskScheduler.Default awaitable

### How use

```c#

await Avalonia.Threading.Dispatcher.UIThread; // Switch to UI Thered with Normal priority

// Execute instruction on UI Thread with Normal priority

await TaskScheduler.Default;

// Execute instruction on TaskScheduler Thread context

await Avalonia.Threading.Dispatcher.UIThread.AwaitWithPriority(Avalonia.Threading.DispatcherPriority.Background); // Switch to UI Thered with Normal Background

// Execute instruction on UI Thread with Background priority

``` 

> **Note**
> I already have a working prototype

**Comment by maxkatz6** [comment]
I would vote for this WPF API instead: https://github.com/AvaloniaUI/Avalonia/issues/14199

**Reply by workgroupengineering** [reply]
Hi,
I didn't know `Dispatcher.Yield`. I looked into, it is a bit different from my proposal.

`Dispatcher.Yield` is similar to VB6 `DoEvents()`, it allows you to wait for the Dispatcher queue to be consumed.

My proposal is to allow switching from non-UI threads to UI thread and vice versa.
